session/pingpong/invoice_tracker.go

```diff
@@ -0,0 +1,176 @@
+	exchangeMessageChan chan crypto.ExchangeMessage,
+	exchangeMessageWaitTimeout time.Duration) *InvoiceTracker {
+	return &InvoiceTracker{
+		peer:                           peer,
+		stop:                           make(chan struct{}),
+		peerInvoiceSender:              peerInvoiceSender,
+		exchangeMessageChan:            exchangeMessageChan,
+		exchangeMessageWaitTimeout:     exchangeMessageWaitTimeout,
+		chargePeriod:                   chargePeriod,
+		maxNotReceivedExchangeMessages: calculateMaxNotReceivedExchangeMessageCount(chargePeriodLeeway, chargePeriod),
+	}
+}
+
+func calculateMaxNotReceivedExchangeMessageCount(chargeLeeway, chargePeriod time.Duration) uint64 {
+	return uint64(math.Round(float64(chargeLeeway) / float64(chargePeriod)))
+}
+
+// Start stars the invoice tracker
+func (it *InvoiceTracker) Start() error {
+	log.Debug("Starting...")
+	// give the consumer a second to start up his payments before sending the first request
+	firstSend := time.After(time.Second)
+	for {
+		select {
+		case <-firstSend:
+			err := it.sendInvoiceExpectExchangeMessage()
+			if err != nil {
+				return err
+			}
+		case <-it.stop:
+			return nil
+		case <-time.After(it.chargePeriod):
+			err := it.sendInvoiceExpectExchangeMessage()
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (it *InvoiceTracker) markExchangeMessageNotReceived() {
+	atomic.AddUint64(&it.notReceivedExchangeMessageCount, 1)
+}
+
+func (it *InvoiceTracker) resetNotReceivedExchangeMessageCount() {
+	atomic.SwapUint64(&it.notReceivedExchangeMessageCount, 0)
+}
+
+func (it *InvoiceTracker) getNotReceivedExchangeMessageCount() uint64 {
+	return atomic.LoadUint64(&it.notReceivedExchangeMessageCount)
+}
+
+func (it *InvoiceTracker) sendInvoiceExpectExchangeMessage() error {
+	err := it.sendInvoice()
+	if err != nil {
+		return err
+	}
+
+	err = it.receiveExchangeMessageOrTimeout()
+	if err != nil {
+		handlerErr := it.handleExchangeMessageReceiveError(err)
+		if handlerErr != nil {
+			return err
+		}
+	} else {
+		it.resetNotReceivedExchangeMessageCount()
+	}
+	return nil
+}
+
+func (it *InvoiceTracker) handleExchangeMessageReceiveError(err error) error {
+	// if it's a timeout, we'll want to ignore it if we're not exceeding maxNotReceivedexchangeMessages
+	if err == ErrExchangeWaitTimeout {
+		it.markExchangeMessageNotReceived()
+		if it.getNotReceivedExchangeMessageCount() >= it.maxNotReceivedExchangeMessages {
+			return err
+		}
+		log.Warn("Failed to receive exchangeMessage: ", err)
+		return nil
+	}
+	return err
+}
+
+func (it *InvoiceTracker) sendInvoice() error {
+	// TODO: a ton of actions should go here
+
+	// TODO: fill the fields
+	return it.peerInvoiceSender.Send(crypto.Invoice{AgreementID: 1234})
+}
+
+func (it *InvoiceTracker) receiveExchangeMessageOrTimeout() error {
+	select {
+	case pm := <-it.exchangeMessageChan:
+		if res := pm.ValidateExchangeMessage(common.HexToAddress(it.peer.Address)); !res {
+			return ErrExchangeValidationFailed
+		}
```

Comment:
If ValidateExchangeMessage returns boolean when it's common to name variable ok so you know what this is a boolean result.

```ts
if ok := pm.ValidateExchangeMessage(common.HexToAddress(it.peer.Address)); !ok {
    return ErrExchangeValidationFailed
}
```

Reply:
https://github.com/mysteriumnetwork/payments/issues/34